hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCDCLogger.java

```diff
@@ -0,0 +1,263 @@
+  }
+
+  public boolean isEmpty() {
+    return !this.cdcEnabled || this.cdcData.isEmpty();
+  }
+
+  public Option<AppendResult> writeCDCData() {
+    if (isEmpty()) {
+      return Option.empty();
+    }
+    try {
+      Map<HoodieLogBlock.HeaderMetadataType, String> header = buildCDCBlockHeader();
+      List<IndexedRecord> records = cdcData.values().stream()
+          .map(record -> {
+            try {
+              return record.getInsertValue(cdcSchema).get();
+            } catch (IOException e) {
+              throw new HoodieIOException("Failed to get cdc record", e);
+            }
+          }).collect(Collectors.toList());
+      HoodieLogBlock block = new HoodieCDCDataBlock(records, header, keyField);
+      AppendResult result = cdcWriter.appendBlocks(Collections.singletonList(block));
+
+      // call close to trigger the data flush.
+      this.close();
+
+      return Option.of(result);
+    } catch (Exception e) {
+      throw new HoodieException("Failed to write the cdc data to " + cdcWriter.getLogFile().getPath(), e);
+    }
+  }
+
+  private Map<HoodieLogBlock.HeaderMetadataType, String> buildCDCBlockHeader() {
+    Map<HoodieLogBlock.HeaderMetadataType, String> header = new HashMap<>();
+    header.put(HoodieLogBlock.HeaderMetadataType.INSTANT_TIME, commitTime);
+    if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
+      header.put(HoodieLogBlock.HeaderMetadataType.SCHEMA, HoodieCDCUtils.CDC_SCHEMA_STRING);
+    } else if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE)) {
+      header.put(HoodieLogBlock.HeaderMetadataType.SCHEMA,
+          HoodieCDCUtils.CDC_SCHEMA_OP_RECORDKEY_BEFORE_STRING);
+    } else {
+      header.put(HoodieLogBlock.HeaderMetadataType.SCHEMA,
+          HoodieCDCUtils.CDC_SCHEMA_OP_AND_RECORDKEY_STRING);
+    }
+    return header;
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (cdcWriter != null) {
+        cdcWriter.close();
+      }
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to close HoodieCDCLogger", e);
+    } finally {
+      cdcData.clear();
+    }
+  }
+
+  public static void setCDCStatIfNeeded(HoodieWriteStat stat,
+                                        String partitionPath,
+                                        HoodieCDCLogger cdcLogger,
+                                        long recordsWritten,
+                                        long insertRecordsWritten,
+                                        FileSystem fs) {
+    try {
+      if (cdcLogger == null || recordsWritten == 0L || (recordsWritten == insertRecordsWritten)) {
+        // the following cases where we do not need to write out the cdc file:
+        // case 1: all the data from the previous file slice are deleted. and no new data is inserted;
+        // case 2: all the data are new-coming,
+        return;
+      }
+
+      Option<AppendResult> cdcResult = cdcLogger.writeCDCData();
```

Comment:
Seems like my previous comment got displaced somehow:

Let's extract invocation of `writeCDCData` into the caller's method. This method should be scoped to only set CDC info w/in the `HoodieWriteStat`

Reply:
it seems there is a gap between your comment and @xushiyan and @prasannarajaperumal 's . 
see : https://github.com/apache/hudi/pull/6697#discussion_r973884446 and https://github.com/apache/hudi/pull/6697#discussion_r973949629

Reply:
@YannByron i don't think there's any gap: @prasannarajaperumal was referring to moving the method into CDCLogger, and @xushiyan was suggesting to abstract it w/in the static method.

What i was saying is that `setCDCStat` should only be setting up `HoodieWriteStat` -- it shouldn't be calling `writeCDCData`

Reply:
I don't think there's any contradiction b/w these comments:
 - @prasannarajaperumal was suggesting to encapsulate HoodieWriteStat setting w/in HoodieCDCLogger
 - @xushiyan was talking about the same

My comment is about splitting out `writeCDCData` out of `setWriteStat`  method (which should just be setting `HoodieWriteStat`)

Reply:
yes, i understood. so i split two methods in this pr. i will move `writeCDCData` into WriteHandle` in next pr.